base/shell/explorer/trayprop.cpp

```diff
@@ -21,441 +21,240 @@
+            DeleteObject(hbmpTaskbar);
+        if (hbmpTray)
+            DeleteObject(hbmpTray);
+    }
+
+    LRESULT OnInitDialog(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+    {
+        CheckDlgButton(IDC_TASKBARPROP_LOCK, g_TaskbarSettings.bLock ? BST_CHECKED : BST_UNCHECKED);
+        CheckDlgButton(IDC_TASKBARPROP_HIDE, g_TaskbarSettings.sr.AutoHide ? BST_CHECKED : BST_UNCHECKED);
+        CheckDlgButton(IDC_TASKBARPROP_ONTOP, g_TaskbarSettings.sr.AlwaysOnTop ? BST_CHECKED : BST_UNCHECKED);
+        CheckDlgButton(IDC_TASKBARPROP_GROUP, g_TaskbarSettings.bGroupButtons ? BST_CHECKED : BST_UNCHECKED);
+        //CheckDlgButton(IDC_TASKBARPROP_SHOWQL, g_TaskbarSettings.bShowQuickLaunch ? BST_CHECKED : BST_UNCHECKED);
+        CheckDlgButton(IDC_TASKBARPROP_CLOCK, (!g_TaskbarSettings.sr.HideClock) ? BST_CHECKED : BST_UNCHECKED);
+        CheckDlgButton(IDC_TASKBARPROP_SECONDS, g_TaskbarSettings.bShowSeconds ? BST_CHECKED : BST_UNCHECKED);
+        CheckDlgButton(IDC_TASKBARPROP_HIDEICONS, g_TaskbarSettings.bHideInactiveIcons ? BST_CHECKED : BST_UNCHECKED);
+
+        UpdateBitmaps();
+        return TRUE;
+    }
+
+    LRESULT OnCtrlCommand(WORD wNotifyCode, WORD wID, HWND hWndCtl, BOOL &bHandled)
+    {
+        UpdateBitmaps();
+        SetModified(TRUE);
+        return 0;
+    }
 
-    if (hwndLock && hwndHide && hwndGroup && hwndShowQL && hwndClock && hwndSeconds && hwndHideInactive)
+    int OnApply()
     {
-        bLock = (SendMessage(hwndLock, BM_GETCHECK, 0, 0) == BST_CHECKED);
-        bHide = (SendMessage(hwndHide, BM_GETCHECK, 0, 0) == BST_CHECKED);
-        bGroup = (SendMessage(hwndGroup, BM_GETCHECK, 0, 0) == BST_CHECKED);
-        bShowQL = (SendMessage(hwndShowQL, BM_GETCHECK, 0, 0) == BST_CHECKED);
+        TaskbarSettings *newSettings = new TaskbarSettings();
+        memcpy(newSettings, &g_TaskbarSettings, sizeof(TaskbarSettings));
+
+        newSettings->bLock = IsDlgButtonChecked(IDC_TASKBARPROP_LOCK);
+        newSettings->sr.AutoHide = IsDlgButtonChecked(IDC_TASKBARPROP_HIDE);
+        newSettings->sr.AlwaysOnTop = IsDlgButtonChecked(IDC_TASKBARPROP_ONTOP);
+        newSettings->bGroupButtons = IsDlgButtonChecked(IDC_TASKBARPROP_GROUP);
+        //newSettings->bShowQuickLaunch = IsDlgButtonChecked(IDC_TASKBARPROP_SHOWQL);
+        newSettings->sr.HideClock = !IsDlgButtonChecked(IDC_TASKBARPROP_CLOCK);
+        newSettings->bShowSeconds = IsDlgButtonChecked(IDC_TASKBARPROP_SECONDS);
+        newSettings->bHideInactiveIcons = IsDlgButtonChecked(IDC_TASKBARPROP_HIDEICONS);
+        
+        newSettings->Save();
+        SendMessage(hwndTaskbar, TWM_SETTINGSCHANGED, 0, (LPARAM)newSettings);
         
-        bShowClock = (SendMessage(hwndClock, BM_GETCHECK, 0, 0) == BST_CHECKED);
-        bShowSeconds = (SendMessage(hwndSeconds, BM_GETCHECK, 0, 0) == BST_CHECKED);
-        bHideInactive = (SendMessage(hwndHideInactive, BM_GETCHECK, 0, 0) == BST_CHECKED);
+        return PSNRET_NOERROR;
```

Comment:
It didn't leak, the receipient of TWM_SETTINGSCHANGED freed it. However I already changed it locally to send a pointer from the sender's stack (and this is fine because SendMessage blocks when sent to a different thread).